### ISSUE TITLE
New AbstractASECalculator abstract type

### ISSUE THREAD

**Issue by jameskermode** [libAtoms/JuLIP.jl#29]
I would like to be able to create multiple subtypes of ASE Calculator, e.g. to allow dispatching differently for DFT-based calculators.

How about renaming the current `ASECalculator` to `AbstractASECalculator` and adding the following:

`ASECalculator <: AbstractASECalculator <: AbstractCalculator`
`GPAWCalculator <: AbstractASECalculator <: AbstractCalculator`

**Comment by cortner** [comment]
`ASEPotential <: ASECalculator <: AbstractCalculator`
`GPAWCalculator <: ASEQuantumCalculator <: ASECalculator` ????

this would only make sense if there is substantial interface shared between different DFT bindings?

**Comment by jameskermode** [comment]
Good idea- there is likely to be quite a bit of shared functionality between different DFT interfaces. There is a DFTCalculator base class in ASE, would be good to use same name:

https://wiki.fysik.dtu.dk/ase/ase/calculators/calculators.html

**Comment by cortner** [comment]
Should we create a module `DFTCalculators` with interface definitions?

**Comment by jameskermode** [comment]
Yup, sounds good. I can have a go at this next week.

**Comment by cortner** [comment]
That would be great - thank you. So just to confirm: 
```
module DFT
abstract AbstractDFTCalculator <: Calculator 
abstract ASEDFTCalculator <: ASECalculator
typealias DFTCalculator Union{AbstractDFTCalculator, ASEDFTCalculator}
type GPAWCalculator <: ASEDFTCalculator
end 
end 
```

This is the best I can come up with to deal with the issue that we don't have multiple inheritance.

Alternatively we can just assume that all DFT calculators will come via ASE, but there is at least one [DFT code](https://github.com/vonDonnerstein/QuantumLab.jl) already being developed in Julia so it might be nice to stay general?

**Comment by jameskermode** [comment]
I agree with keeping `AbstractDFTCalculator` in the hierarchy.

It's nicer than I thought, actually, I didn't realise we could have ``Abstract <: Concrete <: Abstract``, rather than ``Concrete <: Abstract <: Abstract``, so I was expecting to have to rename ``ASECalculator`` to ``AbstractASECalculator`` and then add a new concrete ``ASECalculator``.

**Comment by cortner** [comment]
I actually missed that - sorry - I don't think we can have `Concrete <: Abstract <: Abstract`.  Only leaf-types can be concrete.  :(

**Comment by jameskermode** [comment]
OK, that was what I expected, so I will add the `AbstractASECalculator` type.

**Comment by cortner** [comment]
How about ASEPotential <: ASECalculator ?

**Comment by cortner** [comment]
(I.e. make ASECalculator abstract)

**Comment by jameskermode** [comment]
I thought we made a semantic distinction between potentials and calculators: potentials have forces which are the total derivative of some Hamiltonian, while calculators may not?

**Comment by cortner** [comment]
I am starting to get confused; does this mean we should do 
```
    ASEDFTCalculator  <: DFT <: Potential <: AbstractCalculator
    ASENotaPotential <: AbstractCalculator
    typealias ASECalculator = Union{ASEPotential, ASENotaPotential}
    typealias DFTCalculator Union{AbstractDFTCalculator, ASEDFTCalculator}
```
:) 

Maybe for now we should not worry about clean type hierarchies and just define something minimal to make things work? If this ever becomes and issue we can return to it.

**Comment by cortner** [comment]
P.S.: we could start using [traits](https://github.com/mauro3/SimpleTraits.jl).

**Comment by cortner** [comment]
see also [Traitor.jl](https://github.com/andyferris/Traitor.jl)

**Comment by jameskermode** [comment]
Agreed, let's find something minimal that works. I just meant that making a distinction between an abstract ASECalculator and a concrete ASEPotential was a bit of a hack. How about


```julia
module ASE
abstract AbstractASECalculator <: Calculator
type ASECalculator <: AbstractASECalculator
end

module DFT
abstract AbstractDFTCalculator <: Calculator 
typealias AbstactASEDFTCalculator Union{AbstractDFTCalculator, AbstractASECalculator}
type GPAWCalculator <: AbstractASEDFTCalculator
end
```

**Comment by cortner** [comment]
that type hierarchy won't work - I don't think you can make a subtype of a type-union. Let's put that also on our list of things to settle in person.

**Comment by jameskermode** [comment]
Agreed to implement `AbstractASECalculator`

**Comment by jameskermode** [comment]
Closed by #40
